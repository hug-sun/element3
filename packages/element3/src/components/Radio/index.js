import ElRadio from './src/Radio.vue'

/* istanbul ignore next */
ElRadio.install = function (app) {
  app.component(ElRadio.name, ElRadio)
}

export { ElRadio }
