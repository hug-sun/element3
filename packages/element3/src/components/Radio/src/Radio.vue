<template>
  <label
    role="radio"
    class="el-radio"
    :class="labelClass"
    :aria-checked="radioValue === label"
    :aria-disabled="isDisabled"
    :tabindex="tabIndex"
  >
    <span
      class="el-radio__input"
      :class="{
        'is-disabled': isDisabled,
        'is-checked': radioValue === label
      }"
    >
      <span class="el-radio__inner"></span>
      <input
        type="radio"
        class="el-radio__original"
        :value="label"
        v-model="radioValue"
        :name="name"
        aria-hidden="true"
        :disabled="isDisabled"
        @focus="focus = true"
        @blur="focus = false"
        tabindex="-1"
      />
    </span>
    <span class="el-radio__label" @keydown.stop>
      <slot>{{ label }}</slot>
    </span>
  </label>
</template>

<script>
import { ref, computed, inject, getCurrentInstance, defineComponent } from 'vue'
import { props } from './props'

export default defineComponent({
  name: 'ElRadio',
  componentName: 'ElRadio',
  props,
  emits: ['update:modelValue', 'update:value', 'change'],
  setup(props, context) {
    const focus = ref(false)
    const { elForm, elFormItem } = useInject()
    const { isGroup, radioGroup } = useCheckGroup()
    const radioValue = computed({
      get: () => (isGroup ? radioGroup.proxy.modelValue : props.modelValue),
      set: (value) => {
        changeHandler(value)
      }
    })
    const { isDisabled, radioSize, tabIndex } = useStyle({
      props,
      isGroup,
      radioGroup,
      elForm,
      elFormItem,
      radioValue
    })

    const labelClass = useLabelClass({
      props,
      radioSize,
      radioValue,
      isDisabled,
      focus
    })
    const changeHandler = (value) => {
      context.emit('update:modelValue', value)
      isGroup && radioGroup.emit('update:modelValue', value)
      context.emit('change', value)
      isGroup && radioGroup.emit('change', value)
    }
    return {
      focus,
      radioValue,
      isDisabled,
      radioSize,
      tabIndex,
      labelClass,
      changeHandler
    }
  }
})
const useInject = () => {
  const elForm = inject('elForm', {})
  const elFormItem = inject('elFormItem', {})
  return { elForm, elFormItem }
}
const useCheckGroup = () => {
  const { parent } = getCurrentInstance()
  const isGroup = parent.type.name === 'ElRadioGroup'
  const radioGroup = isGroup ? parent : null
  return { isGroup, radioGroup }
}
const useStyle = ({
  props,
  isGroup,
  radioGroup,
  elForm,
  elFormItem,
  radioValue
}) => {
  const {
    proxy,
    parent: {
      proxy: { radioGroupSize }
    }
  } = getCurrentInstance()
  const elFormDisable = elForm.disabled
  const isDisabled = computed(() =>
    isGroup
      ? radioGroup.props.disabled || props.disabled || elFormDisable
      : props.disabled || elFormDisable
  )

  const radioSize = computed(
    () =>
      props.size ||
      radioGroupSize ||
      (elForm && elFormItem.elFormItemSize) ||
      (proxy.$ELEMENT || {}).size
  )
  const tabIndex = computed(() =>
    isDisabled.value || (isGroup && radioValue.value !== props.label) ? -1 : 0
  )

  return { isDisabled, radioSize, tabIndex }
}
const useLabelClass = ({ props, radioSize, radioValue, isDisabled, focus }) =>
  computed(() => [
    props.border && radioSize.value ? `el-radio--${radioSize.value}` : '',
    { 'is-checked': radioValue.value === props.label },
    { 'is-disabled': isDisabled.value },
    { 'is-focus': focus.value },
    { 'is-bordered': props.border }
  ])
</script>

<style></style>
