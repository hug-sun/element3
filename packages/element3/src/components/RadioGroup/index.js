import ElRadioGroup from './src/RadioGroup'

ElRadioGroup.install = function (app) {
  app.component(ElRadioGroup.name, ElRadioGroup)
}

export { ElRadioGroup }
