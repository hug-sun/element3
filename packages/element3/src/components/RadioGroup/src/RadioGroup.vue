<template>
  <div class="el-radio-group" role="radiogroup">
    <slot></slot>
  </div>
</template>

<script lang="ts">
import { props } from './props'
import { defineComponent, inject, computed, toRefs, watch } from 'vue'
import { useGlobalOptions } from '../../../composables/globalConfig'
import { useEmitter } from '../../../composables/emitter'

export default defineComponent({
  name: 'ElRadioGroup',

  props,

  emits: ['update:modelValue', 'change'],

  setup(props) {
    const { size, modelValue } = toRefs(props)
    const globalConfig = useGlobalOptions()
    const elFormItem = inject('elFormItem', {})
    const { dispatch } = useEmitter()

    watch(modelValue, (v) => {
      dispatch('el.form.change', v)
    })

    const radioGroupSize = useRadioGroupSize({ size, elFormItem, globalConfig })

    return {
      radioGroupSize
    }
  }
})

const useRadioGroupSize = ({ size, elFormItem, globalConfig }) => {
  return computed(() => {
    return (
      size?.value || (elFormItem as any)?.elFormItemSize || globalConfig.size
    )
  })
}
</script>
