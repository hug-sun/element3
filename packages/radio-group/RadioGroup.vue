<template @change="test">
  <component
    :is="elTag"
    class="el-radio-group"
    role="radiogroup"
    @keydown="handleKeydown"
  >
    <slot></slot>
  </component>
</template>
<script>
import { useEmitter } from '../../src/use/emitter'
import {
  getCurrentInstance,
  onMounted,
  watch,
  toRefs,
  computed,
  inject
} from 'vue'
const keyCode = Object.freeze({
  LEFT: 37,
  UP: 38,
  RIGHT: 39,
  DOWN: 40
})

export default {
  name: 'ElRadioGroup',

  componentName: 'ElRadioGroup',

  props: {
    modelValue: [String, Number, Symbol, Boolean],
    size: String,
    fill: String,
    textColor: String,
    disabled: Boolean
  },

  emits: ['update:modelValue', 'change'],

  setup(props) {
    const { modelValue = {}, size = {} } = toRefs(props)
    const { dispatch } = useEmitter()
    const { elFormItem } = useInject()

    watch(modelValue, (v) => {
      dispatch('el.form.change', v)
    })

    const { handleKeydown } = useKeyDown()

    const { elTag, radioGroupSize } = useStyle({ elFormItem, size })

    return {
      elTag,
      radioGroupSize,
      handleKeydown
    }
  }
}

function useInject() {
  const elFormItem = inject('elFormItem', {})
  return {
    elFormItem
  }
}

function useKeyDown() {
  const { ctx } = getCurrentInstance()
  onMounted(() => {
    const radios = ctx.$el.querySelectorAll('[type=radio]')
    const firstLabel = ctx.$el.querySelectorAll('[role=radio]')[0]
    if (![].some.call(radios, (radio) => radio.checked) && firstLabel) {
      firstLabel.tabIndex = 0
    }
  })

  function handleKeydown(e) {
    const target = e.target
    const className =
      target.nodeName === 'INPUT' ? '[type=radio]' : '[role=radio]'
    const radios = ctx.$el.querySelectorAll(className)
    const length = radios.length
    const index = [].indexOf.call(radios, target)
    const roleRadios = ctx.$el.querySelectorAll('[role=radio]')
    switch (e.keyCode) {
      case keyCode.LEFT:
      case keyCode.UP:
        e.stopPropagation()
        e.preventDefault()
        if (index === 0) {
          roleRadios[length - 1].click()
          roleRadios[length - 1].focus()
        } else {
          roleRadios[index - 1].click()
          roleRadios[index - 1].focus()
        }
        break
      case keyCode.RIGHT:
      case keyCode.DOWN:
        if (index === length - 1) {
          e.stopPropagation()
          e.preventDefault()
          roleRadios[0].click()
          roleRadios[0].focus()
        } else {
          roleRadios[index + 1].click()
          roleRadios[index + 1].focus()
        }
        break
      default:
        break
    }
  }

  return { handleKeydown }
}

function useStyle({ elFormItem, size }) {
  const { ctx, vnode } = getCurrentInstance()

  const radioGroupSize = computed(() => {
    return size.value || elFormItem.elFormItemSize || (ctx.$ELEMENT || {}).size
  })

  const elTag = computed(() => {
    return (vnode.data || {}).tag || 'div'
  })

  return { radioGroupSize, elTag }
}
</script>
