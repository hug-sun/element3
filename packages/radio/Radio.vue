<template>
  <label
    class="el-radio"
    :class="[
      border && radioSize ? 'el-radio--' + radioSize : '',
      { 'is-disabled': isDisabled },
      { 'is-focus': focus },
      { 'is-bordered': border },
      { 'is-checked': model === label }
    ]"
    role="radio"
    :aria-checked="model === label"
    :aria-disabled="isDisabled"
    :tabindex="tabIndex"
    @keydown.space.stop.prevent="model = isDisabled ? model : label"
  >
    <span
      class="el-radio__input"
      :class="{
        'is-disabled': isDisabled,
        'is-checked': model === label
      }"
    >
      <span class="el-radio__inner"></span>
      <input
        ref="radio"
        class="el-radio__original"
        :value="label"
        type="radio"
        aria-hidden="true"
        v-model="model"
        @focus="focus = true"
        @blur="focus = false"
        @change="handleChange"
        :name="name"
        :disabled="isDisabled"
        tabindex="-1"
      />
    </span>
    <span class="el-radio__label" @keydown.stop>
      <slot></slot>
      <template v-if="!$slots.default">{{ label }}</template>
    </span>
  </label>
</template>
<script>
import {
  ref,
  computed,
  toRefs,
  nextTick,
  getCurrentInstance,
  inject,
  unref
} from 'vue'

export default {
  name: 'ElRadio',

  componentName: 'ElRadio',

  props: {
    modelValue: [String, Number, Symbol, Boolean],
    label: [String, Number, Symbol, Boolean],
    disabled: Boolean,
    name: String,
    border: Boolean,
    size: String
  },

  emits: ['update:modelValue', 'change'],

  setup(props) {
    const { modelValue, label, size, disabled } = toRefs(props)
    const radio = ref()
    const focus = ref(false)

    const { elForm, elFormItem } = useInject()

    const { isGroup, radioGroup } = useCheckGroup()

    const { model, handleChange } = useModel({
      isGroup,
      radioGroup,
      modelValue,
      label,
      radio
    })

    const { radioSize, isDisabled, tabIndex } = useStyle({
      isGroup,
      radioGroup,
      size,
      disabled,
      model,
      label,
      elForm,
      elFormItem
    })
    return {
      radio,
      focus,
      model,
      radioSize,
      isDisabled,
      tabIndex,
      handleChange
    }
  }
}

function useInject() {
  const elForm = inject('elForm', {})
  const elFormItem = inject('elFormItem', {})
  return {
    elForm,
    elFormItem
  }
}

function useCheckGroup() {
  let { parent } = getCurrentInstance()
  while (parent) {
    if (parent.type.name !== 'ElRadioGroup') {
      parent = parent.parent
    } else {
      return {
        isGroup: true,
        radioGroup: parent
      }
    }
  }
  return {
    isGroup: false,
    radioGroup: null
  }
}

function useModel({ isGroup, radioGroup, modelValue, radio, label }) {
  const { emit } = getCurrentInstance()

  const model = computed({
    get() {
      const res = isGroup ? radioGroup.proxy.modelValue : modelValue
      return unref(res)
    },
    set(val) {
      if (isGroup) {
        radioGroup.emit('update:modelValue', val)
      } else {
        emit('update:modelValue', val)
      }
      radio.value && (radio.value.checked = unref(model) === unref(label))
    }
  })

  async function handleChange() {
    await nextTick()
    emit('change', model.value)
    isGroup && radioGroup.emit('change', model.value)
  }
  return { model, handleChange }
}

function useStyle({
  isGroup,
  radioGroup,
  size,
  disabled,
  model,
  label,
  elForm,
  elFormItem
}) {
  const { ctx } = getCurrentInstance()
  const elFormDisable = elForm.disabled
  const radioSize = computed(() => {
    const temRadioSize =
      unref(size) || elFormItem.elFormItemSize || (ctx.$ELEMENT || {}).size
    return isGroup
      ? radioGroup.ctx.radioGroupSize || temRadioSize
      : temRadioSize
  })
  const isDisabled = computed(() => {
    return isGroup
      ? radioGroup.props.disabled || disabled.value || elFormDisable
      : disabled.value || elFormDisable
  })

  const tabIndex = computed(() => {
    return isDisabled.value || (isGroup && model.value !== label.value) ? -1 : 0
  })

  return {
    radioSize,
    isDisabled,
    tabIndex
  }
}
</script>
