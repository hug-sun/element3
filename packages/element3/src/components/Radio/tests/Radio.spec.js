import { mount } from '@vue/test-utils'
import { ref, h, reactive } from 'vue'
import RadioGroup from '../../RadioGroup/src/RadioGroup.vue'
import Radio from '../src/Radio.vue'

describe('Radio.vue', () => {
  it('snapshot', () => {
    const wrapper = mount(Radio)
    expect(wrapper.element).toMatchSnapshot()
  })
  describe('props', () => {
    it('disabled', () => {
      const wrapper = mount(Radio, {
        props: {
          disabled: true
        }
      })
      expect(wrapper.classes()).toContain('is-disabled')
      expect(wrapper.vm.isDisabled).toEqual(true)
    })
    it('border and size', () => {
      const wrapper = mount(Radio, {
        props: {
          border: true,
          size: 'small'
        }
      })
      expect(wrapper.vm.radioSize).toEqual('small')
      expect(wrapper.classes()).toContain('is-bordered')
      expect(wrapper.classes()).toContain('el-radio--small')
    })
    it('v-model', async () => {
      const radioVal = ref(0)
      const wrapper = mount(Radio, {
        props: {
          label: 1,
          modelValue: radioVal,
          'onUpdate:modelValue': (val) => {
            radioVal.value = val
          }
        }
      })
      expect(wrapper.vm.radioValue).toBe(0)
      await wrapper.trigger('click')
      expect(wrapper.vm.radioValue).toBe(1)
    })
    it('by elForm.disable', () => {
      const wrapper = mount(Radio, {
        global: {
          provide: {
            elForm: reactive({
              disabled: true
            })
          }
        }
      })
      expect(wrapper.classes()).toContain(`is-disabled`)
    })
    it('set props event change', async () => {
      const radioVal = ref(0)
      const wrapper = mount(Radio, {
        props: {
          label: 1,
          modelValue: radioVal,
          'onUpdate:modelValue': (val) => {
            radioVal.value = val
          },
          disabled: true
        }
      })
      expect(wrapper.classes()).toContain('is-disabled')
      expect(wrapper.vm.radioValue).toBe(0)
      await wrapper.trigger('click')
      expect(wrapper.vm.radioValue).not.toBe(1)
    })
  })
})

describe('Radio Group', () => {
  describe('props', () => {
    it('disabled and size', () => {
      const radioVal = ref(1)
      const wrapper = mount(RadioGroup, {
        props: {
          modelValue: radioVal,
          'onUpdate:modelValue': (val) => {
            radioVal.value = val
          },
          disabled: true
        },
        slots: {
          default: [1, 2, 3].map((item) =>
            h(Radio, {
              label: item
            })
          )
        }
      })
      expect(wrapper.findAllComponents('.is-disabled').length).toEqual(3)
    })
    it('set single disabled to change event', async () => {
      const radioVal = ref(1)
      const wrapper = mount(RadioGroup, {
        props: {
          modelValue: radioVal,
          'onUpdate:modelValue': (val) => {
            radioVal.value = val
          }
        },
        slots: {
          default: [1, 2, 3].map((item) =>
            h(Radio, {
              label: item,
              disabled: item === 2
            })
          )
        }
      })
      expect(wrapper.vm.modelValue).toBe(1)
      expect(wrapper.findAllComponents(Radio).length).toBe(3)
      const radioTem = wrapper.findAllComponents(Radio)
      expect(radioTem[1].classes()).toContain('is-disabled')
      await radioTem[1].trigger('click')
      expect(wrapper.vm.modelValue).not.toBe(2)
      await radioTem[2].trigger('click')
      expect(wrapper.vm.modelValue).toBe(3)
    })
  })
})
