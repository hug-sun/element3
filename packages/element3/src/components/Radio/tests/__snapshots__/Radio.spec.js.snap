// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio.vue snapshot 1`] = `
<label
  aria-checked="true"
  class="el-radio is-checked"
  role="radio"
  tabindex="0"
>
  <span
    class="el-radio__input is-checked"
  >
    <span
      class="el-radio__inner"
    />
    <input
      aria-hidden="true"
      class="el-radio__original"
      name=""
      tabindex="-1"
      type="radio"
      value=""
    />
  </span>
  <span
    class="el-radio__label"
  >
    
    
    
  </span>
</label>
`;
